app/views/Project/Tagging/Sources/BulkUpload/index.tsx

```diff
@@ -1,65 +1,339 @@
+                        ), 'leads');
+                    }
+                }
+            },
+            onError: (gqlError) => {
+                alert.show(
+                    'Failed to save leads!',
+                    { variant: 'error' },
+                );
+                // eslint-disable-next-line no-console
+                console.error(gqlError);
+            },
+        },
+    );
+
+    const handleLeadChange = useCallback(
+        (val: SetValueArg<PartialLeadType>, otherName: number | undefined) => {
+            onLeadChange(
+                (oldValue) => {
+                    const newVal = !isCallable(val)
+                        ? val
+                        : val(oldValue);
+                    return newVal;
+                },
+                otherName,
+            );
+        },
+        [onLeadChange],
+    );
 
     const handleFileUploadSuccess = useCallback((value: FileUploadResponse) => {
         setUploadedFiles((oldUploadedFiles) => ([
             value,
             ...oldUploadedFiles,
         ]));
-    }, []);
-
-    const handleDeleteFile = useCallback((id: number) => {
-        setUploadedFiles((oldState: FileUploadResponse[]) => {
-            const updatedState = produce(oldState, (safeState) => {
-                const index = safeState.findIndex((file: FileUploadResponse) => file.id === id);
-                if (index !== -1) {
-                    safeState.splice(index, 1);
-                }
+        const newLead: PartialLeadType = {
+            clientId: randomString(),
+            sourceType: sourceTypeMap[value.sourceType],
+            confidentiality: 'UNPROTECTED',
+            priority: 'LOW',
+            isAssessmentLead: false,
+            attachment: String(value.id),
+            title: value.title,
+        };
+        setFormFieldValue(
+            (oldVal: PartialFormType['leads']) => [
+                ...(oldVal ?? []),
+                newLead,
+            ],
+            'leads',
+        );
+        if (!selectedLead) {
+            setSelectedLead(newLead.clientId);
+        }
+    }, [setUploadedFiles, setFormFieldValue, selectedLead]);
+
+    const handleLeadRemove = useCallback((clientId: string) => {
+        const selectedLeadObj = formValue?.leads?.find((lead) => lead.clientId === clientId);
+        setFormFieldValue((oldVal) => oldVal?.filter((lead) => lead.clientId !== clientId), 'leads');
+        if (selectedLeadObj?.attachment) {
+            setUploadedFiles((oldState: FileUploadResponse[]) => {
+                const updatedState = produce(oldState, (safeState) => {
+                    const index = safeState.findIndex(
+                        (file) => String(file.id) === selectedLeadObj.attachment,
+                    );
+                    if (index !== -1) {
+                        safeState.splice(index, 1);
+                    }
+                });
+                return updatedState;
```

Comment:
Are we cleaning up the uploaded files from memory?
Is this also done after the leads are saved and all of the leads are cleared?

Reply:
Also, do we need to do this?